[Enhancement] Pause Owl Warp

mckinlee opened the pull request:
**Overview**

This enhancement allows players to select owl warp locations directly from the pause menu map. This feature eliminates the need to play the Song of Soaring to warp, except when in dungeons. In dungeons, players still need to play the song to warp back to the entrance.

**Index Warping Mode**

Index Warping Mode is a special feature that activates if the player has only the hidden owl statue (owl statue 15) activated. In this mode, players can warp based on locations they have visited rather than activated owl statues. However, activating any real owl warp statue will disable Index Warping Mode.

**Index Warping Map**

Visiting the following locations unlocks these owl warps:

- Great Bay to Great Bay Coast
- Zora Hall to Zora Cape
- Romani Ranch to Snowhead
- Deku Palace to Mountain Village
- Woodfall to Clock Town
- Clock Town to Milk Road
- Snowhead to Woodfall
- Ikana Graveyard to Southern Swamp
- Ikana Canyon to Ikana Canyon
- Goron Village to Stone Tower

**Requirements**

Players must have the Song of Soaring and at least one owl statue activated for this enhancement to work. For Index Warping Mode, players are still required to have the Song of Soaring, but are also required to have **_ONLY_** the hidden owl statue activated. If any other statue is activated, Index Warping Mode is disabled.

~~This PR introduces the initial implementation of the enhancement. This feature allows players to warp from the pause menu map rather than having to play the Song of Soaring. The current state of the enhancement is incomplete, and this PR aims to gather feedback for further development.~~

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1772624507.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1772626351.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1772627812.zip)
<!--- section:artifacts:end -->

inspectredc commented on mm/src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_scope_NES.c:
```diff
@@ -3998,6 +4002,17 @@ void KaleidoScope_Update(PlayState* play) {
                     Audio_PlaySfx_PauseMenuOpenOrClose(SFX_PAUSE_MENU_CLOSE);
                     play->msgCtx.ocarinaMode = sCursorPointsToOcarinaModes[pauseCtx->cursorPoint[PAUSE_WORLD_MAP]];
                     Audio_PlaySfx(NA_SE_SY_DECIDE);
+                    // Check if the pause warp enhancement is enabled. If so, and the last played song is not the Song of Soaring,
+                    // teleport the player to the location specified by the cursor point on the world map.
+                    // The reason for the Song of Soaring check is that if the player played the song, the normal logic
+                    // in the codebase will handle the teleport and animations, so manual teleportation is not needed.
+                    if (CVarGetInteger("gEnhancements.Songs.PauseWarp", 0)) {
+                        if (play->msgCtx.lastPlayedSong != OCARINA_SONG_SOARING) {
+                            play->nextEntrance = D_80AF343C[pauseCtx->cursorPoint[PAUSE_WORLD_MAP]];
+                            play->transitionTrigger = TRANS_TRIGGER_START;
+                            play->transitionType = TRANS_TYPE_FADE_WHITE;
+                        }
+                    }
```
I'm wondering if it may just be easier to set the lastPlayedSong as song of soaring? This way it doesn't necessarily need to be placed here (for consolodation under a single hook)

mckinlee replied:
So I've looked into this further and still have not come up with a way to do this unfortunately. Setting lastPlayedSong alone is just not enough to trigger the logic. Gonna leave it as unless someone has a suggestion. :/

mckinlee commented:
Moved back to draft. I messed it up lol

mckinlee committed:
Add Pause Owl Warp Enhancement

- **mm/2s2h/BenGui/BenMenuBar.cpp**: Updated menu to include new enhancement options.
- **mm/2s2h/Enhancements/Enhancements.cpp**: Registered the new Pause Owl Warp enhancement.
- **mm/2s2h/Enhancements/Enhancements.h**: Declared functions for the new enhancement.
- **mm/2s2h/Enhancements/Songs/PauseOwlWarp.cpp**: Implemented the Pause Owl Warp functionality.
- **mm/2s2h/Enhancements/Songs/PauseOwlWarp.h**: Header file for the Pause Owl Warp implementation.
- **mm/include/functions.h**: Added function declarations related to the new enhancement.
- **mm/src/overlays/kaleido_scope/ovl_kaleido_map.c**: Modified to support the new enhancement.
- **mm/src/overlays/kaleido_scope/ovl_kaleido_scope_NES.c**: Modified to support the new enhancement.

mckinlee committed:
Merge branch 'HarbourMasters:develop' into pausewarpredo

mckinlee committed:
ran clang-format script

mckinlee commented:
This has been rewritten from the ground up and is now in a state ready for review. :)

mckinlee committed:
Additional logic and bug fixes

mckinlee committed:
Fix PauseOwlWarp cursor logic triggering on dungeon map
-Added a check to ensure PauseOwlWarp's cursor logic is only triggered when not in a dungeon.

mckinlee committed:
Add Song of Soaring check
- To make it more in line with vanilla, you are now required to have the Song of Soaring to use this enhancement.

mckinlee committed:
bug fix and ran clang-format script
-My previous commit had a extra condition that wasn't needed that caused a bug where the world map points locations pointed to activated owl statues.

mckinlee committed:
added more checks

mckinlee committed:
Simplified complex conditional statements

mckinlee committed:
possibly fix macos build issue?

mckinlee committed:
Merge branch 'HarbourMasters:develop' into pausewarp

mckinlee committed:
index warping logic added

Archez commented on mm/2s2h/Enhancements/Songs/PauseOwlWarp.cpp:
```diff
@@ -0,0 +1,220 @@
+extern u16 D_80AF343C[];
+extern s16 sInDungeonScene;
+
+bool IsOwlWarpEnabled() {
```
For style matching with the project, I recommend moving this function out of this `extern "C" { }` section, and giving it its own extern C keyword in front. And I recommend giving it a more verbose name since this func is considered external/exported, something like:

`extern "C" bool PauseOwlWarp_IsOwlWarpEnabled() {`

Archez commented on mm/2s2h/Enhancements/Songs/PauseOwlWarp.h:
```diff
@@ -0,0 +1,14 @@
+extern "C" {
+#endif
+
+void RegisterPauseOwlWarp();
```
You can add the declaration for `IsOwlWarpEnabled` in here so that it is externed via the header file.

Archez commented on mm/src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_map.c:
```diff
@@ -13,6 +13,8 @@
 
 #include "BenPort.h"
 
+extern bool IsOwlWarpEnabled();
```
Then you can remove this extern in favor of including your header `#include "2s2h/Enhancements/Songs/PauseOwlWarp.h"`

Archez commented on mm/src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_scope_NES.c:
```diff
@@ -24,6 +24,8 @@
 
 #include "2s2h/Enhancements/GameInteractor/GameInteractor.h"
 
+extern bool IsOwlWarpEnabled();
```
Same thing here

Archez commented on mm/2s2h/Enhancements/Songs/PauseOwlWarp.cpp:
```diff
@@ -0,0 +1,220 @@
+bool IsOwlWarpEnabled() {
+    return CVarGetInteger("gEnhancements.Songs.PauseOwlWarp", 0) && CHECK_QUEST_ITEM(QUEST_SONG_SOARING) &&
+           (gSaveContext.save.saveInfo.playerData.owlActivationFlags != 0 ||
+            gSaveContext.save.saveInfo.playerData.owlActivationFlags == (1 << 15));
```
Could this just be simplified to be `!= 0`, since we just care that _any_ statue is activated?

Archez commented on mm/include/functions.h:
```diff
@@ -902,6 +902,8 @@ void KaleidoScopeCall_Init(PlayState* play);
 void KaleidoScopeCall_Update(PlayState* play);
 void KaleidoScopeCall_Draw(PlayState* play);
+void KaleidoScope_UpdateOwlWarpNamePanel(PlayState* play);
+void KaleidoScope_UpdateNamePanel(PlayState* play);
```
Move these two down  at the bottom of the file in the `#region 2S2H [Port]` section for "Previously unavailable functions, made available for porting".

Archez commented on mm/2s2h/Enhancements/Songs/PauseOwlWarp.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+static bool isConfirming = false;
+static u16 sStickAdjTimer = 0;
```
Just like the stick timer one, could you change `isConfirming` to `sIsConfirming` to signify that the variable is a file level static.

Archez commented on mm/2s2h/Enhancements/Songs/PauseOwlWarp.cpp:
```diff
@@ -0,0 +1,220 @@
+
+void RegisterPauseOwlWarp() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnKaleidoUpdate>([](PauseContext* pauseCtx) {
+        if (!sInDungeonScene && IsOwlWarpEnabled() && CHECK_QUEST_ITEM(QUEST_SONG_SOARING)) {
```
This hook body is larger enough that you could move the logic to a dedicated function, and just call into it from here.

Archez reviewed (commented):
First off, great work on this! I love to see how encapsulated all this logic is, and initial playtesting looks and feels solid as well.

I've included some initial feedback below, mostly around code styling.

mckinlee committed:
implement feedback from Archez

mckinlee commented:
> First off, great work on this! I love to see how encapsulated all this logic is, and initial playtesting looks and feels solid as well.
> 
> I've included some initial feedback below, mostly around code styling.



> First off, great work on this! I love to see how encapsulated all this logic is, and initial playtesting looks and feels solid as well.
> 
> I've included some initial feedback below, mostly around code styling.

Thank you for the feedback! I have made those changes you recommended. However, I'm not sure why the OTR failed to generate over here, when it worked just fine on my fork. Might need to rerun it.

mckinlee committed:
shot in the dark macos build fix?

mckinlee committed:
how bout now?

mckinlee committed:
Merge remote-tracking branch 'upstream/develop' into pausewarp

mckinlee committed:
revert back to previous commit

mckinlee committed:
if this fixed the macos build issue it'll be thanks to cclark25... let's pray

mckinlee committed:
add debugeditor check - thanks archez!

mckinlee committed:
Moved debugEditor check to prevent triggering ALL pauseOwlWarp logic. wtf was I thinking earlier
